File: scripts/decoding/i2c.py
#!/usr/bin/env python3
"""I²C decoder (SCL clock + SDA data).

START = SDA↓ while SCL high; STOP = SDA↑ while SCL high; bits sampled on SCL rising edges, MSB first,
8 data + 1 ACK per byte. The first byte after a START is the 7-bit address + R/W. Self-framing, so a
gapless (continuous) stream needs no bit-grid tricks — but a capture window that catches no START has
no boundary to lock onto. Run `python3 -m decoding.i2c` to self-test."""
import numpy as np


def decode_i2c(scl, sda):
    """Decode an I²C bus. Returns START/STOP markers plus byte dicts
    [{start, end, value, ack, text, kind}] (kind 'addr' for the first byte after START)."""
    scl = np.asarray(scl).astype(np.int8)
    sda = np.asarray(sda).astype(np.int8)
    n = len(scl)
    events = []
    bits, byte_start = [], None
    in_frame = False
    expect_addr = False
    for i in range(1, n):
        if scl[i] == 1 and scl[i - 1] == 1:
            if sda[i - 1] == 1 and sda[i] == 0:              # START
                events.append({'start': i, 'end': i, 'text': 'S', 'kind': 'start'})
                bits, byte_start = [], None
                in_frame, expect_addr = True, True
                continue
            if sda[i - 1] == 0 and sda[i] == 1:              # STOP
                events.append({'start': i, 'end': i, 'text': 'P', 'kind': 'stop'})
                bits, byte_start, in_frame = [], None, False
                continue
        if in_frame and scl[i] == 1 and scl[i - 1] == 0:     # sample on SCL rising
            if byte_start is None:
                byte_start = i
            bits.append(sda[i])
            if len(bits) == 9:
                val = 0
                for b in bits[:8]:
                    val = (val << 1) | b
                ack = (bits[8] == 0)
                if expect_addr:
                    rw = 'R' if (val & 1) else 'W'
                    text = f"{val >> 1:02X}{rw}{'A' if ack else 'N'}"; kind = 'addr'
                    expect_addr = False
                else:
                    text = f"{val:02X}{'A' if ack else 'N'}"; kind = 'byte'
                events.append({'start': byte_start, 'end': i, 'value': val,
                               'ack': ack, 'text': text, 'kind': kind})
                bits, byte_start = [], None
    return events


# --- self-test -------------------------------------------------------------------
def _synth_i2c(values, spb=10, addr=0x50, rw=0):
    """SCL+SDA: START, addr+RW (+ACK), each data byte (+ACK), STOP."""
    scl, sda = [], []

    def seg(c, d, k):
        scl.extend([c] * k); sda.extend([d] * k)

    seg(1, 1, spb); seg(1, 0, spb)                          # idle, then START
    for byte in [(addr << 1) | rw] + list(values):
        for b in range(7, -1, -1):
            bit = (byte >> b) & 1
            seg(0, bit, spb // 2); seg(1, bit, spb); seg(0, bit, spb // 2)
        seg(0, 0, spb // 2); seg(1, 0, spb); seg(0, 0, spb // 2)    # ACK
    seg(0, 0, spb // 2); seg(1, 0, spb // 2); seg(1, 1, spb)        # STOP
    return np.array(scl, dtype=int), np.array(sda, dtype=int)


def selftest():
    ramp = list(range(256))
    scl, sda = _synth_i2c(ramp, addr=0x50, rw=0)
    ev = decode_i2c(scl, sda)
    data = [e['value'] for e in ev if e['kind'] == 'byte']
    good = (sum(e['kind'] == 'start' for e in ev) == 1 and sum(e['kind'] == 'stop' for e in ev) == 1
            and data == ramp and all(e['ack'] for e in ev if e['kind'] in ('addr', 'byte')))
    print(f"I2C  addr+ramp: {len(data)} data bytes, {'OK' if good else 'FAIL'}")
    return good


if __name__ == '__main__':
    import sys
    sys.exit(0 if selftest() else 1)

File: backend/src/decoder/i2c.rs
//! I²C decoder — SCL clock plus SDA data.
//!
//! START is SDA falling while SCL is high, STOP is SDA rising while SCL is high, and bits
//! are sampled on SCL rising edges, MSB first, 8 data bits plus an ACK. The first byte
//! after a START carries the address and R/W.
//!
//! The bus is self-framing, so a gapless stream needs none of the bit-grid machinery UART
//! requires — but a capture window that catches no START has no boundary to lock onto,
//! which is what the end-anchored fallback exists for.

use super::{Event, Kind};

/// How byte boundaries are established.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum Anchor {
    /// Forward only — bytes counted from each START.
    Start,
    /// Forward when a START was captured, else count backward from the transaction end so
    /// a capture that missed the START still recovers its bytes.
    #[default]
    Auto,
    /// Force the end-anchored fallback.
    End,
}

/// Decode an I²C bus.
pub fn decode(scl: &[bool], sda: &[bool], anchor: Anchor) -> Vec<Event> {
    let events = decode_forward(scl, sda);
    let has_start = events.iter().any(|e| e.kind == Kind::Start);
    match anchor {
        Anchor::Start => events,
        Anchor::Auto if has_start => events,
        _ => decode_end_anchored(scl, sda, events),
    }
}

/// Forward decode: bits counted from each START.
///
/// Handles repeated START, 7-bit and 10-bit addressing, and ACK/NACK.
fn decode_forward(scl: &[bool], sda: &[bool]) -> Vec<Event> {
    let n = scl.len().min(sda.len());
    let mut events = Vec::new();
    let mut bits: Vec<bool> = Vec::with_capacity(9);
    let mut byte_start: Option<usize> = None;
    let mut in_frame = false;
    let mut expect_address = false;
    let mut expect_address_low = false;

    for i in 1..n {
        // START and STOP are the two SDA transitions that occur while SCL is held high.
        if scl[i] && scl[i - 1] {
            if sda[i - 1] && !sda[i] {
                events.push(Event {
                    start: i,
                    end: i,
                    value: None,
                    ok: true,
                    kind: if in_frame { Kind::RepeatedStart } else { Kind::Start },
                });
                bits.clear();
                byte_start = None;
                in_frame = true;
                expect_address = true;
                expect_address_low = false;
                continue;
            }
            if !sda[i - 1] && sda[i] {
                events.push(Event {
                    start: i,
                    end: i,
                    value: None,
                    ok: true,
                    kind: Kind::Stop,
                });
                bits.clear();
                byte_start = None;
                in_frame = false;
                expect_address_low = false;
                continue;
            }
        }

        if in_frame && scl[i] && !scl[i - 1] {
            byte_start.get_or_insert(i);
            bits.push(sda[i]);
            if bits.len() == 9 {
                let value = bits[..8].iter().fold(0u8, |acc, &b| (acc << 1) | u8::from(b));
                let ack = !bits[8];
                let kind = if expect_address && (value & 0xF8) == 0xF0 {
                    // 10-bit address, first byte: 11110xx + R/W.
                    expect_address = false;
                    expect_address_low = true;
                    Kind::Address
                } else if expect_address {
                    expect_address = false;
                    Kind::Address
                } else if expect_address_low {
                    expect_address_low = false;
                    Kind::Address
                } else {
                    Kind::Byte
                };
                events.push(Event {
                    start: byte_start.unwrap_or(i),
                    end: i,
                    value: Some(value),
                    ok: ack,
                    kind,
                });
                bits.clear();
                byte_start = None;
            }
        }
    }
    events
}

/// Fallback for a capture with no START, triggered mid-transaction.
///
/// Byte boundaries cannot be counted from a START, so they are anchored to the transaction
/// end instead: SCL rising edges are grouped into 9-clock bytes counting backward from the
/// last STOP, dropping the leading partial. Bytes come out as plain data because the
/// address is off-screen. The direct analog of SPI end-anchoring.
fn decode_end_anchored(scl: &[bool], sda: &[bool], events: Vec<Event>) -> Vec<Event> {
    let n = scl.len().min(sda.len());
    let rises: Vec<usize> = (1..n).filter(|&i| scl[i] && !scl[i - 1]).collect();
    let falls: Vec<usize> = (1..n).filter(|&i| !scl[i] && scl[i - 1]).collect();

    let last_stop = events
        .iter()
        .filter(|e| e.kind == Kind::Stop)
        .map(|e| e.start)
        .next_back();

    let usable: Vec<usize> = match last_stop {
        Some(stop) => {
            // A STOP releases SDA with SCL held high, adding a rising edge that is not a
            // data or ACK bit. The last falling edge before the STOP ends the real final
            // clock, so only rises before that count.
            let cutoff = falls.iter().copied().rfind(|&f| f < stop).unwrap_or(stop);
            rises.into_iter().filter(|&r| r < cutoff).collect()
        }
        None => rises,
    };
    if usable.len() < 9 {
        return events;
    }
    let usable = &usable[usable.len() % 9..];

    let mut out: Vec<Event> = events.into_iter().filter(|e| e.kind == Kind::Stop).collect();
    for group in usable.chunks_exact(9) {
        let value = group[..8]
            .iter()
            .fold(0u8, |acc, &i| (acc << 1) | u8::from(sda[i]));
        out.push(Event {
            start: group[0],
            end: group[8],
            value: Some(value),
            ok: !sda[group[8]],
            kind: Kind::Byte,
        });
    }
    out.sort_by_key(|e| e.start);
    out
}
